tests/tree_inference/test_simulate.py

```diff
@@ -1 +1,276 @@
+            homozygous_rate,
+        ]
+    )
+
+    # Generate the matrix of y values
+    perfect_mat = random.choice(
+        rng, jnp.arange(len(probabilities)), shape=shape, p=probabilities
+    )
+
+    return perfect_mat
+
+
+@pytest.mark.parametrize("seed,", [42])
+@pytest.mark.parametrize("false_positive_rate,", [1e-5, 0.1])
+@pytest.mark.parametrize("false_negative_rate,", [1e-2, 0.3])
+@pytest.mark.parametrize("missing_entry_rate,", [0.0, 1e-2, 0.3, 0.5])
+@pytest.mark.parametrize("observe_homozygous,", [True, False])
+def test_na_freq(
+    seed: int,
+    false_positive_rate: float,
+    false_negative_rate: float,
+    missing_entry_rate: float,
+    observe_homozygous: bool,
+):
+    """test for expected frequency of NAs in noisy mutation matrix"""
+
+    # RNGs for false positives, false negatives, and missing data
+    rng = random.PRNGKey(seed)
+    n = 300
+    m = 300
+    shape = (n, m)
+    pos_rate = 0.3
+    homozygous_rate = 0.0
+    if observe_homozygous:
+        homozygous_rate = 0.1
+
+    perfect_mat = perfect_matrix(rng, pos_rate, homozygous_rate, shape)
+
+    # generate noisy matrices
+    noisy_mat = sim.add_noise_to_perfect_matrix(
+        rng,
+        perfect_mat,
+        false_positive_rate,
+        false_negative_rate,
+        missing_entry_rate,
+        observe_homozygous,
+    )
+
+    freq_na = np.sum(noisy_mat == 3) / (n * m)
+
+    # three standard deviations - stdev = (N * p * (1-p))^0.5
+    tolerance = 3 * ((n * m) * missing_entry_rate * (1 - missing_entry_rate)) ** 0.5
+    tolerance_freq = tolerance / (n * m)
+
+    assert pytest.approx(missing_entry_rate, abs=tolerance_freq) == freq_na
+
+
+@pytest.mark.parametrize("seed,", [42])
+@pytest.mark.parametrize("false_positive_rate,", [1e-5, 0.1, 0.5])
+@pytest.mark.parametrize("false_negative_rate,", [1e-2, 0.3])
+@pytest.mark.parametrize("missing_entry_rate,", [1e-2])
+@pytest.mark.parametrize("observe_homozygous,", [True, False])
+def test_fp(
+    seed: int,
+    false_positive_rate: float,
+    false_negative_rate: float,
+    missing_entry_rate: float,
+    observe_homozygous: bool,
+):
+    """test for expected frequency of false positives in noisy mutation matrix"""
+
+    # RNGs for false positives, false negatives, and missing data
+    rng = random.PRNGKey(seed)
+    n = 300
+    m = 300
+    shape = (n, m)
+    pos_rate = 0.3
+    homozygous_rate = 0.0
+    if observe_homozygous:
+        homozygous_rate = 0.1
+    neg_rate = 1 - pos_rate - homozygous_rate
+
+    perfect_mat = perfect_matrix(rng, pos_rate, homozygous_rate, shape)
+
+    # generate noisy matrices
+    noisy_mat = sim.add_noise_to_perfect_matrix(
+        rng,
+        perfect_mat,
+        false_positive_rate,
+        false_negative_rate,
+        missing_entry_rate,
+        observe_homozygous,
+    )
+
+    freq_fp = np.sum((noisy_mat == 1) & (perfect_mat == 0)) / (n * m)
+    #
+    # # # three standard deviations - stdev = (N * p * (1-p))^0.5
+    # tolerance = (
```

Comment:
The next six lines look like commented out code, so it'd be good to remove it.

(Comments should contain only human-readable information + instructions for the code linters. Commented-out code gets outdated pretty quickly.)

Reply:
OK removed - makes sense. This was my attempt to find a statistically meaningful tolerance. It was too rigorous for some reason, hence I reverted to absolute fixed tolerance of 3% frequency.